crates/ruff/src/noqa.rs

```diff
@@ -255,41 +273,202 @@ enum ParsedFileExemption<'a> {
 
 impl<'a> ParsedFileExemption<'a> {
     /// Return a [`ParsedFileExemption`] for a given comment line.
-    fn try_extract(line: &'a str) -> Option<Self> {
-        let line = line.trim_whitespace_start();
+    fn try_extract(line: &'a str) -> Result<Option<Self>, ParseError> {
+        let line = ParsedFileExemption::lex_whitespace(line);
+        let Some(line) = ParsedFileExemption::lex_char(line, '#') else {
+            return Ok(None);
+        };
+        let line = ParsedFileExemption::lex_whitespace(line);
+
+        if let Some(line) = ParsedFileExemption::lex_flake8(line) {
+            // Ex) `# flake8: noqa`
+            let line = ParsedFileExemption::lex_whitespace(line);
+            let Some(line) = ParsedFileExemption::lex_char(line, ':') else {
+                return Ok(None);
+            };
```

Comment:
@MichaReiser - Is there any more succinct or idiomatic way to express this? I.e., propagate the `None`?

Reply:
Not that I know. I find it readable.